core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/api/rename.kt

```diff
@@ -0,0 +1,46 @@
+package org.jetbrains.kotlinx.dataframe.impl.api
+
+import org.jetbrains.kotlinx.dataframe.DataFrame
+import org.jetbrains.kotlinx.dataframe.api.RenameClause
+import org.jetbrains.kotlinx.dataframe.api.cast
+import org.jetbrains.kotlinx.dataframe.api.getColumnsWithPaths
+import org.jetbrains.kotlinx.dataframe.api.insert
+import org.jetbrains.kotlinx.dataframe.api.under
+import org.jetbrains.kotlinx.dataframe.columns.ColumnKind
+import org.jetbrains.kotlinx.dataframe.columns.ColumnWithPath
+import org.jetbrains.kotlinx.dataframe.impl.columns.tree.allChildrenNotNull
+import org.jetbrains.kotlinx.dataframe.impl.columns.tree.collectTree
+import org.jetbrains.kotlinx.dataframe.kind
+
+internal fun <T, C> RenameClause<T, C>.renameImpl(newNames: Array<out String>): DataFrame<T> {
+    var i = 0
+    return renameImpl { newNames[i++] }
+}
+
+internal fun <T, C> RenameClause<T, C>.renameImpl(transform: (ColumnWithPath<C>) -> String): DataFrame<T> {
+    val selectedColumns = df.getColumnsWithPaths(columns)
+    val tree = df.getColumnsWithPaths { all().rec() }.collectTree()
+
+    // perform rename in nodes
+    tree.allChildrenNotNull().forEach { node ->
+        val column = selectedColumns.find { it.data == node.data } ?: return@forEach
+        val newName = transform(column)
+        node.name = newName
+    }
+
+    // build up a new DataFrame using the modified names
+    var newDf = DataFrame.empty(df.rowsCount()).cast<T>()
+    tree.allChildrenNotNull().forEach { node ->
```

Comment:
This tree structure is used in other places to create dataframes? Is there an existing code to re-create a df from it? Approach with insert under looks suspicious because it's a "mutation" of an immutable data structure in a loop

Reply:
There doesn't appear to be such a thing. Usually dataframes are built from the ground up with columns recursively. I'll try to create a sort-of mapping function for TreeNode with which I can create a new dataframe with columns I create on the fly